### ISSUE TITLE
reproduction of HAMMER results

### ISSUE THREAD

**Issue by Daniellli** [ByteDance-Seed/manip-as-in-sim-suite#13]
good work!
I am planning to reproduce the quantitative results reported in your paper. Is it possible to release the eval code?
Best regards

**Comment by Ericonaldo** [comment]
Yeah, code for you to refer to:

```
    def compute_metrics(self, output, batch):
        B = batch["image"].shape[0]
        metrics_dict = {}
        for b in range(B):
            if self._compute_abs_metric or self._compute_rel_metric:
                pred_depth = output["depth"][b][0].detach().cpu().numpy()
                lowres_depth = batch["lowres_depth"][b][0].detach().cpu().numpy()
                if "mesh_depth" in batch:
                    Log.info("Using mesh depth for evaluation", tag="mesh_depth")
                    gt_depth = batch["mesh_depth"][b][0].detach().cpu().numpy()
                else:
                    gt_depth = batch["depth"][b][0].detach().cpu().numpy()
                msk = self.create_depth_mask(batch["dataset_name"], gt_depth)

            if self._compute_abs_metric:
                metrics_dict_item = self.compute_depth_metric_bin(pred_depth, gt_depth, msk)
                metrics_dict = self.update_metrics_dict(
                    metrics_dict, metrics_dict_item, "absolute"
                )

                pred_depth = self.align_depth_func(
                    pred_depth,
                    lowres_depth,
                    msk,
                    # disp=("disp" in output and output["disp"]),
                )
                metrics_dict_item = self.compute_depth_metric(pred_depth, gt_depth, msk)
                metrics_dict = self.update_metrics_dict(
                    metrics_dict, metrics_dict_item, "absolute_prompt_align"
                )

            if self._compute_rel_metric:
                pred_depth = self.align_depth_func(
                    pred_depth,
                    gt_depth,
                    msk,
                    # disp=("disp" in output and output["disp"]),
                )
                metrics_dict_item = self.compute_depth_metric(pred_depth, gt_depth, msk)
                metrics_dict = self.update_metrics_dict(
                    metrics_dict, metrics_dict_item, "relative"
                )
        return metrics_dict

    def update_metrics_dict(self, metrics_dict, metrics_dict_item, prefix):
        for k, v in metrics_dict_item.items():
            if f"{prefix}_{k}" not in metrics_dict:
                metrics_dict[f"{prefix}_{k}"] = []
            metrics_dict[f"{prefix}_{k}"].append(v)
        return metrics_dict

    def create_depth_mask(self, dataset_name, gt_depth):
        # return gt_depth > 1e-3 
        return np.logical_and(gt_depth > 1e-3, ~np.isnan(gt_depth)) & (~np.isinf(gt_depth))
        
    def compute_depth_metric_bin(self, pred_depth, gt_depth, msk):
        gt = gt_depth[msk]
        pred = pred_depth[msk]
        thresh = np.maximum((gt / (pred + 1e-5)), (pred / (gt + 1e-5)))
        t11 = (thresh < 1.1).mean()
        d05 = (thresh < 1.25**0.5).mean()
        d1 = (thresh < 1.25).mean()
        d2 = (thresh < 1.25**2).mean()
        d3 = (thresh < 1.25**3).mean()

        l1 = np.mean(np.abs(gt - pred))

        rmse = (gt - pred) ** 2
        rmse = np.sqrt(rmse.mean())
        
        abs_rel = np.mean(np.abs(gt - pred) / (gt + 1e-5))
        sq_rel = np.mean(((gt - pred) ** 2) / (gt + 1e-5))

        metrics =  {
            "t11": t11,
            "d0.5": d05,
            "d1": d1,
            "d2": d2,
            "d3": d3,
            "abs_rel": abs_rel,
            "sq_rel": sq_rel,
            "rmse": rmse,
            "l1": l1,
        }

        num_hist, max_distance = 300, 3.0

        # 计算距离区间统计
        distance_bins = np.linspace(0, max_distance, num_hist + 1)
        distance_centers = (distance_bins[:-1] + distance_bins[1:]) / 2

        # 初始化统计数组
        pixel_counts = np.zeros(num_hist)
        abs_rel_errors = np.full(num_hist, np.nan)
        l1_errors = np.full(num_hist, np.nan)
        error_percent = np.full(num_hist, np.nan)

        # 对每个距离区间进行统计
        for i in range(num_hist-1):
            # 找到当前距离区间内的像素
            mask_in_bin = (gt >= distance_bins[i]) & (gt < distance_bins[i + 1])

            count_in_bin = np.count_nonzero(mask_in_bin)
            pixel_counts[i] = count_in_bin

            if count_in_bin > 0:
                gt_in_bin = gt[mask_in_bin]
                pred_in_bin = pred[mask_in_bin]

                # 计算该区间内的误差
                abs_rel_errors[i] = np.mean(np.abs(gt_in_bin - pred_in_bin) / (gt_in_bin + 1e-5)) 
                l1_errors[i] = np.mean(np.abs(gt_in_bin - pred_in_bin))
                error_percent[i] = np.mean(np.abs(gt_in_bin-pred_in_bin) / (gt_in_bin + 1e-5))

            metrics[f"pixel_counts_{distance_centers[i]}-{distance_centers[i+1]}_meter"] = pixel_counts[i]
            metrics[f"abs_rel_{distance_centers[i]}-{distance_centers[i+1]}_meter"] = abs_rel_errors[i] * pixel_counts[i] if not np.isnan(abs_rel_errors[i]) else 0.
            metrics[f"l1_{distance_centers[i]}-{distance_centers[i+1]}_meter"] = l1_errors[i] * pixel_counts[i] if not np.isnan(l1_errors[i]) else 0.
            metrics[f"error_percent_{distance_centers[i]}-{distance_centers[i+1]}_meter"] = error_percent[i] if not np.isnan(error_percent[i]) else 0.
        
        return metrics

    def compute_depth_metric(self, pred_depth, gt_depth, msk):
        gt = gt_depth[msk]
        pred = pred_depth[msk]
        thresh = np.maximum((gt / (pred + 1e-5)), (pred / (gt + 1e-5)))
        t11 = (thresh < 1.1).mean()
        d05 = (thresh < 1.25**0.5).mean()
        d1 = (thresh < 1.25).mean()
        d2 = (thresh < 1.25**2).mean()
        d3 = (thresh < 1.25**3).mean()

        l1 = np.mean(np.abs(gt - pred))

        rmse = (gt - pred) ** 2
        rmse = np.sqrt(rmse.mean())
        
        abs_rel = np.mean(np.abs(gt - pred) / (gt + 1e-5))
        sq_rel = np.mean(((gt - pred) ** 2) / (gt + 1e-5))

        return {
            "t11": t11,
            "d0.5": d05,
            "d1": d1,
            "d2": d2,
            "d3": d3,
            "abs_rel": abs_rel,
            "sq_rel": sq_rel,
            "rmse": rmse,
            "l1": l1,
        }
```
